source/dub/recipe/packagerecipe.d

```diff
@@ -171,50 +213,139 @@ struct ConfigurationInfo {
 	}
 }
 
+/// Actual representation of a dependency as specified by the user
+private struct DependencyObject
+{
+	@Name("version")
+	@Optional string version_;
+	@Optional string path;
+	@Optional string repository;
+	bool optional;
+	@Name("default")
+	bool default_;
+
+	@Name("dflags", true) string[][string] dflags;
+
+	public void validate () const
+	{
+		enforce(this.optional || !this.default_,
+				"Setting default to `true` has no effect if `optional` is not set");
+		enforce(this.version_.length || this.path.length || this.repository.length,
+				"Need to provide one of the following fields: `version`, `path`, or `repository`");
+
+		enforce(!this.path.length || !this.repository.length,
+				"Cannot provide a `path` dependency if a repository dependency is used");
+		enforce(!this.repository.length || this.version_.length,
+				"Need to provide a commit as `version` with repository dependency");
+
+		// Need to deprecate this as it's fairly common
+		version (none) {
+			enforce(!this.path.length || !this.version_.length,
+					"Cannot provide a `path` dependency if a `version` dependency is used");
```

Comment:
where is this used? path + version doesn't seem to make much sense imo

Reply:
Usually it's `version="*" path="...."`. Example: https://github.com/dlang/dub/issues/2252

Reply:
this seems like a bug, version should probably already be implicitly be set to `*` with a path based dependency? Can just add a comment in that issue about it though and we can worry about that in a separate PR.

Reply:
Yes that's why this is `version (none)`. Re-working the config file parsing allowed me to find all those little issues and my approach is to fix them if it can be done easily, or be compatible with them and make a note about it, like here.